### ISSUE TITLE
Automatic conversion of ""platform"" code

### ISSUE THREAD

**Issue by benjamingr** [domenic/get-originals#14]
I think there is a lot of value in prototyping a tool that converts "regular" code not using originals into code that is "safe" and using originals.

Ideally, we would use such a tool in Node.js to enforce https://github.com/nodejs/node/issues/18795 and the web platform can use it to safely write APIs in JavaScript and expose them to the window.

This would require optimisation from the engine in order to be viable for platforms but this is definitely interesting to us.

Once the semantics here are finalised (let me know) I would love to take a stab at this and possibly build such a tool under the Node.js foundation. (cc @hashseed)

**Comment by domenic** [comment]
Yeah, we're actively exploring building such tooling in the Chrome team too. I jotted down some initial notes at https://github.com/domenic/get-originals#automatic-rewritingenforcement-of-robustness.

**Comment by ljharb** [comment]
How would you reliably detect usage of prototype methods? For example, ".has" does a brand check on Maps, but it could be called on any object at all, and many objects will have a method called "has".

**Comment by domenic** [comment]
There's a lot of options, and a lot of exploration left to do. Off the top of my head:

- With a type system, you would be able to tell if the LHS of the "." is a Map. This is complicated by the unsoundness of TypeScript as a popular type system; I can't tell if Flow is sound, although I see vague gestures in that direction. But soundness is only necessary if we want 100% perfection, which would be nice but IMO is not necessary if we have good enough heuristics.
- As noted in the above link, one approach would be to just disallow all usage of "." at all. This is over-cautious, but in a generally good way, e.g. it would also prevent you from accidentally using your own API's public methods/properties when you really should be using private state. It remains to be seen whether this would work in practice, but skimming some of the code in async-local-storage, jsdom, and Chrome's Streams implementation, it looks like it might work.
